### ISSUE TITLE
Custom loss function labels and predictions order preservation

### ISSUE THREAD

**Issue by selectasterisk** [dmlc/xgboost#4260]
Suppose 
```
Xs = np.random.rand(5,2)
Ys = [1,1,0,1,1]
dtrain = xgb.DMatrix(Xs, Ys)

xgb_params = {
        'base_score': 0.1,
        'booster': 'gbtree',
        'n_jobs': -1,
        'objective': custom_loss
    }
xgb_clf = xgb.XGBClassifier(**xgb_params)

```

If I am to write a custom objective function using XGBoost (sklearn API):

```
def custom_loss(y, y_hat):
 
    p = y_hat
    grad = p - y
    hess = p * (1 - p)
    return grad, hess
```

Can I expect that the labels to be in the same order every time? (e.g. y == [1,1,0,1,1], and the corresponding predictions, of course)

What if I use `n_jobs=-1`?

Aside question: since 'objective'  is set to the custom function, (and not binary:logistic); is y_hat just the margin response? (e.g. raw predicted output from -inf to inf)?

Also, it seems like after we call xgb_clf.fit, print(xgb_clf.get_params()['objective'] changes.

Code to reproduce:
```
import numpy as np
import xgboost as xgb
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn import datasets

from sklearn.metrics import precision_score,accuracy_score
iris = datasets.load_iris()
X = iris.data
y = iris.target
X=X[y<=1]
y=y[y<=1]

def custom_loss(y, y_hat):
    #y_hat = 1.0 / (1.0 + np.exp(-y_hat))
    p = y_hat
    grad = p - y
    hess = p * (1 - p)
    return grad, hess

xgb_params = {
        'base_score': 0.3,
        'booster': 'gbtree',
        'n_jobs': -1,
        'objective': custom_loss
    }

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# fit model no training data
model = xgb.XGBClassifier(**xgb_params)
print(model.get_params()['objective'])
model.fit(X_train, y_train)
print(model.get_params()['objective'])

# make predictions for test data
y_pred = model.predict(X_test)
predictions = [round(value) for value in y_pred]
# evaluate predictions
accuracy = accuracy_score(y_test, predictions)
print("Accuracy: %.2f%%" % (accuracy * 100.0))
```

**Comment by hcho3** [comment]
@selectasterisk 

> Can I expect that the labels to be in the same order every time?

Yes, the labels will be in the same order every time. The reason is that your `custom_loss` is a vectorized function, so the arguments `y` and `yhat` are already collection of labels.

**Comment by hcho3** [comment]
> is y_hat just the margin response?

Yes, it is just a margin response.

**Comment by selectasterisk** [comment]
Thanks for the response. To follow up on this (for sklearn custom objective) it seems like y_hat on XGBClassifier is actually a post-sigmoid response, and to get the error function to work, all I had to do was to differentiate wrt to the post-sigmoid activation, and not wrt to the margin.
